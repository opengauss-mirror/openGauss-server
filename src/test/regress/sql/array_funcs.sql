select array_remove(array[], 2);
select array_remove(array[1,2,2,3], 2);
select array_remove(array[1,2,2,3], 5);
select array_remove(array[1,NULL,NULL,3], NULL);
select array_remove(array['A','CC','D','C','RR'], 'RR');
select array_remove('{{1,2,2},{1,4,3}}', 2); -- not allowed
select array_remove(array['X','X','X'], 'X') = '{}';
select array_replace(array[[],[]],5,3);
select array_replace(array[1,2,5,4],5,3);
select array_replace(array[1,2,5,4],5,NULL);
select array_replace(array[1,2,NULL,4,NULL],NULL,5);
select array_replace(array['A','B','DD','B'],'B','CC');
select array_replace(array[1,NULL,3],NULL,NULL);
select array_replace(array['AB',NULL,'CDE'],NULL,'12');
select array_replace(array[[NULL,NULL],[NULL,NULL]],NULL,'12');
select array_replace(array[['AB',NULL,'CDE'],['CD','CDE',NULL]],NULL,'12');
select array_replace(array[[[1],[1]],[[1],[1]]], 1, 2);
select array_replace(array[[[1],[1]],[[NULL],[1]]], 1, NULL);

--array test in function, $1 is a string, $2 is delimiter, $3 is index, $4 is mode(for test purpose)
CREATE OR REPLACE FUNCTION substring_index(varchar, varchar, integer, integer)
RETURNS varchar AS $$
  DECLARE
    tokens varchar[];
    length integer ;
    indexnum integer;
  BEGIN
    tokens := pg_catalog.string_to_array($1, $2);
    length := pg_catalog.array_upper(tokens, 1);
    indexnum := length - ($3 * -1) + 1;
    IF $3 >= 0 THEN
      RETURN pg_catalog.array_to_string(tokens[1:$3], $2);
    ELSEIF $4 = 0 THEN
      RETURN pg_catalog.array_to_string(tokens[indexnum:length], $2);
    ELSEIF $4 = 1 THEN
      RETURN pg_catalog.array_to_string(tokens[indexnum:6], $2);
    ELSEIF $4 = 2 THEN
      RETURN pg_catalog.array_to_string(tokens[2:length], $2);
    ELSEIF $4 = 3 THEN
      RETURN pg_catalog.array_to_string(tokens[2:4], $2);
    END IF;
  END;
$$ IMMUTABLE STRICT LANGUAGE PLPGSQL;
create table array_test(a int[7]);
insert into array_test values('{12,3,4,5,6,7,8}');

select substring_index('12,3,4,5,6,7,8', ',', 3, 0), array_to_string(a[1:3], ',') from array_test;
select substring_index('12,3,4,5,6,7,8', ',', -3, 0), array_to_string(a[7 - 3 + 1:7], ',') from array_test;
select substring_index('12,3,4,5,6,7,8', ',', -2, 0), array_to_string(a[7 - 2 + 1:7], ',') from array_test;
select substring_index('12,3,4,5,6,7,8', ',', -3, 1), array_to_string(a[7 - 3 + 1:6], ',') from array_test;
select substring_index('12,3,4,5,6,7,8', ',', -3, 2), array_to_string(a[2:7], ',') from array_test;
select substring_index('12,3,4,5,6,7,8', ',', -3, 3), array_to_string(a[2:4], ',') from array_test;

drop table array_test;
drop function substring_index;
