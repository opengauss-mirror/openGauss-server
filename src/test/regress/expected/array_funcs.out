select array_remove(array[], 2);
ERROR:  cannot determine type of empty array
LINE 1: select array_remove(array[], 2);
                            ^
HINT:  Explicitly cast to the desired type, for example ARRAY[]::integer[].
CONTEXT:  referenced column: array_remove
select array_remove(array[1,2,2,3], 2);
 array_remove 
--------------
 {1,3}
(1 row)

select array_remove(array[1,2,2,3], 5);
 array_remove 
--------------
 {1,2,2,3}
(1 row)

select array_remove(array[1,NULL,NULL,3], NULL);
 array_remove 
--------------
 {1,3}
(1 row)

select array_remove(array['A','CC','D','C','RR'], 'RR');
 array_remove 
--------------
 {A,CC,D,C}
(1 row)

select array_remove('{{1,2,2},{1,4,3}}', 2); -- not allowed
ERROR:  removing elements from multidimensional arrays is not supported
CONTEXT:  referenced column: array_remove
select array_remove(array['X','X','X'], 'X') = '{}';
 ?column? 
----------
 t
(1 row)

select array_replace(array[[],[]],5,3);
ERROR:  cannot determine type of empty array
LINE 1: select array_replace(array[[],[]],5,3);
                                   ^
HINT:  Explicitly cast to the desired type, for example ARRAY[]::integer[].
CONTEXT:  referenced column: array_replace
select array_replace(array[1,2,5,4],5,3);
 array_replace 
---------------
 {1,2,3,4}
(1 row)

select array_replace(array[1,2,5,4],5,NULL);
 array_replace 
---------------
 {1,2,NULL,4}
(1 row)

select array_replace(array[1,2,NULL,4,NULL],NULL,5);
 array_replace 
---------------
 {1,2,5,4,5}
(1 row)

select array_replace(array['A','B','DD','B'],'B','CC');
 array_replace 
---------------
 {A,CC,DD,CC}
(1 row)

select array_replace(array[1,NULL,3],NULL,NULL);
 array_replace 
---------------
 {1,NULL,3}
(1 row)

select array_replace(array['AB',NULL,'CDE'],NULL,'12');
 array_replace 
---------------
 {AB,12,CDE}
(1 row)

select array_replace(array[[NULL,NULL],[NULL,NULL]],NULL,'12');
   array_replace   
-------------------
 {{12,12},{12,12}}
(1 row)

select array_replace(array[['AB',NULL,'CDE'],['CD','CDE',NULL]],NULL,'12');
       array_replace       
---------------------------
 {{AB,12,CDE},{CD,CDE,12}}
(1 row)

select array_replace(array[[[1],[1]],[[1],[1]]], 1, 2);
     array_replace     
-----------------------
 {{{2},{2}},{{2},{2}}}
(1 row)

select array_replace(array[[[1],[1]],[[NULL],[1]]], 1, NULL);
           array_replace           
-----------------------------------
 {{{NULL},{NULL}},{{NULL},{NULL}}}
(1 row)

--array test in function, $1 is a string, $2 is delimiter, $3 is index, $4 is mode(for test purpose)
CREATE OR REPLACE FUNCTION substring_index(varchar, varchar, integer, integer)
RETURNS varchar AS $$
  DECLARE
    tokens varchar[];
    length integer ;
    indexnum integer;
  BEGIN
    tokens := pg_catalog.string_to_array($1, $2);
    length := pg_catalog.array_upper(tokens, 1);
    indexnum := length - ($3 * -1) + 1;
    IF $3 >= 0 THEN
      RETURN pg_catalog.array_to_string(tokens[1:$3], $2);
    ELSEIF $4 = 0 THEN
      RETURN pg_catalog.array_to_string(tokens[indexnum:length], $2);
    ELSEIF $4 = 1 THEN
      RETURN pg_catalog.array_to_string(tokens[indexnum:6], $2);
    ELSEIF $4 = 2 THEN
      RETURN pg_catalog.array_to_string(tokens[2:length], $2);
    ELSEIF $4 = 3 THEN
      RETURN pg_catalog.array_to_string(tokens[2:4], $2);
    END IF;
  END;
$$ IMMUTABLE STRICT LANGUAGE PLPGSQL;
create table array_test(a int[7]);
insert into array_test values('{12,3,4,5,6,7,8}');
select substring_index('12,3,4,5,6,7,8', ',', 3, 0), array_to_string(a[1:3], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 12,3,4          | 12,3,4
(1 row)

select substring_index('12,3,4,5,6,7,8', ',', -3, 0), array_to_string(a[7 - 3 + 1:7], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 6,7,8           | 6,7,8
(1 row)

select substring_index('12,3,4,5,6,7,8', ',', -2, 0), array_to_string(a[7 - 2 + 1:7], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 7,8             | 7,8
(1 row)

select substring_index('12,3,4,5,6,7,8', ',', -3, 1), array_to_string(a[7 - 3 + 1:6], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 6,7             | 6,7
(1 row)

select substring_index('12,3,4,5,6,7,8', ',', -3, 2), array_to_string(a[2:7], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 3,4,5,6,7,8     | 3,4,5,6,7,8
(1 row)

select substring_index('12,3,4,5,6,7,8', ',', -3, 3), array_to_string(a[2:4], ',') from array_test;
 substring_index | array_to_string 
-----------------+-----------------
 3,4,5           | 3,4,5
(1 row)

drop table array_test;
drop function substring_index;
